### PR TITLE
Add parsing from xContent to InternalSearchHit and InternalSearchHits

### PR DESCRIPTION
**Pull request by cbuescher**
In preparation to be able to parse search hits from rest responses for the java rest client, 
this adds methods to parse InternalSearchHit and InternalSearchHits from their
xContent representation. Most of the information in the original object is
preserved when rendering the object to xContent and then parsing it back.
However, some pieces of information are lost which we currently cannot parse
back from the rest response, most notably:

* the "match" property of the lucene explanation is not rendered in the
  "_explain" section and cannot be reconstructed on the client side
* the original "shard" information (SearchShardTarget) is only rendered if the
  "explanation" is also set. However, we always render the "_index" name that 
  is contained in SearchShardTarget. In order to be able to parse back only 
  the index name, this information is now stored separately in InternalSearchHit.
  Also we loose the indexUUID of the contained ShardId because we don't write 
  it out. As a replacement we can use ClusterState.UNKNOWN_UUID on the receiving side.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: Can't we get rid of this and only use SearchShardTarget?

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: nit: shall we make this the only setter (make it public also) and have caller of the other setter for the same member call this instead?

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: shall we have an else block here and throw?

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: not sure why, but it seems like the score can also be null? I don't think we support that here:

```
if (Float.isNaN(score)) {
            builder.nullField(Fields._SCORE);
        } else {
            builder.field(Fields._SCORE, score);
        }
```

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: The problem here is that the response often only contains the "_index" field (other information necessary to build a complete SearchShardTarget are only rendered out under certain conditions, e.g. if explanations are set). So in order to parse the "_index" name back, I either need a place to store it independently of a complete SearchShardTarget, or I need to construct a dummy SearchShardTarget with bogus values. This is the first option, I think its better than to construct a SearchShardTarget full of dummy values.

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: if it's a metadata field it's never a list, I think we can just expect a single value and use parser.objectText(), at least that is what I did in GetResult.

Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: I am wondering if this branch should just be an else, where we can assert or throw if the field name is not a metadata field, because it must really be.

Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: In this case 'list()' wouldn't work I think, because the meta data fields (like '_parent' etc.) are not written as an array but as a single value. Another idea would be to access AbstractXContentParser#readValue() for this, but I remember some related discussion about controlling which token types are parser how that we had for GetField, which is where I took the parseValue() method from. Maybe @javanna has an opinion here as well.

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: I was wondering if the parse value stuff should rather go in InternalSearchHitField, that would become very simular to GetField and we could probably share code between the two as well.

**Comment by cbuescher** [comment]
@tlrx thanks, I addressed some of your comments, left comments on the others where I'd like some more discussion before deciding which way to go.

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: shall we have a common method for this in XContentParserUtils?

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: nit: can we ensure that the current token is START_OBJECT when we enter this method ?

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: Are we sure about this default score value? I am not, but I am not too sure what it should be either

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: as said above, I think it would be nice to share this with GetField given that it's exactly the same code, used to parse stored field values

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: for our own safety, shall we check what these two tokens that we advance are?

File: core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitTests.java
Comment (javanna) [inline core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitTests.java]: `index` is a metadata field and can have length 5 right?. Should we rather pick a higher length, like 10, that no metadata field can reach?

File: core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitTests.java
Comment (javanna) [inline core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitTests.java]: can we realistically not have the version in the hit response?

File: core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitTests.java
Comment (javanna) [inline core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitTests.java]: replace the unknown uuid here too

File: core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitTests.java
Comment (javanna) [inline core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitTests.java]: I don't think this is realistic, in search api it's always going to be a match, never noMatch

File: core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitsTests.java
Comment (javanna) [inline core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitsTests.java]: out of curiosity, why prettyPrint?

**Review by javanna — APPROVED** [review]
I left a bunch of comments but this look very good already.

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: As an aside: Funny how comments may seem out of order when you do a longer review but submit late. I wrote my comment before I read @javannas replies.

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: Yes, I forgot we have direct field access.

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: of course, I started reviewing yesterday evening but only submitted now :)

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: sure

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: This doesn't work here, usually we determine the current field name here and then continue the outer loop to advance the parse pointer. But your point is good, nestedIdentity can be parsed in the START-OBJECT branch, and I think I can change SearchSortValues so it also conforms to how we do it elsewhere.

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: Good catch, I added this case to the test setup and added a branch where null-tokens get parsed to Float.NaN.

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: What did you have in mind? I can make this a static helper method here in InternalSearchHit and use that from GetField, but maybe another place is more appropriate?

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: I think the first token advancing can be in 'fromXContent', the second needs to stay and we don't know if we are on next field name (which we ensure in the next loop) or on closing object token (in which case we exit the loop.

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: I added one.

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: right, they should both be set or not set.

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: I think I will make this a Float so we can validate if both description and value have been parsed before we return the Explanation, otherwise we throw.

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: I would prefer regex over manual string manipulation, I changed it a bit so we also detect the second group is an integer, I think this is better on the long run. I added another test case for strange index names containing brackets (which should be edge cases).

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: not sure, but maybe have InternalSearchHitField fromXContent and toXContent ? Potentially that one and GetField could inherit from the same base abstract class, if that makes sense. Otherwise static helpers would work too. We could also simply add a static helper to XContentParserUtils

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: but the end object is hit only with the above nextToken in the while declaration? I think we could assert just to be sure of what we expect

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: Maybe we can then enforce that, at least then who reads the code understand what the expectations are

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: sounds good!

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: I made this public and reuse it from GetField. Still wondering if this should go to another more common classe though, but I don't have a better ides right now.

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: I don't think we should even worry about edge cases given that we print this out from our own code, but how can a regex be better on the long run? :) it's two concatenated strings, I don't see how a regex helps. It may be me though.

File: core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitTests.java
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitTests.java]: `_index`, `_type` and `_id` are excluded from the metadata field names used here, since this would lead to duplicate json fields (index, type, id is also rendered at another place). I can still make fieldname variable, say 5-10 characters? I just like to keep it readable for debugging purposes.

File: core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitTests.java
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitTests.java]: you are right, its -1 by default. Will delete this.

Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitTests.java]: Thinking about it again, version is settable, so we might want to test what happens if it is not set (in which case the default should be used). Maybe this doesn't happen in out code just yet, but version() setter might not be called after all.

File: core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitTests.java
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitTests.java]: Should I just leave it in in case so we are coverd if somebody start rendering the "match" attribute at some point so we catch that we don't parse it? Maybe not worth it though.

File: core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitsTests.java
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitsTests.java]: Leftover from when I was debugging something. The error messages we get when the xContent output doesn't match has some room for improvement here, but I will remove this for now.

File: core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitTests.java
Comment (javanna) [inline core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitTests.java]: ok then we are good already I think. Leave a comment on why the length is fixed?

File: core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitTests.java
Comment (javanna) [inline core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitTests.java]: that's fine. leave it then

File: core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitTests.java
Comment (javanna) [inline core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitTests.java]: I don't foresee us having that here. The search api can only give explanations about hits, that are implicitly a match. The explain api on the other hand explains also why a certain document doesn't match, in which case it does have a match boolean field. We are good without this.

File: core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitsTests.java
Comment (javanna) [inline core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitsTests.java]: cool

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: > I am wondering if this branch should just be an else, where we can assert or throw

I think I don't understand, this part is in the branch where we parse simple values, if the field name falls through here we throw an exception already. Can you clarify what you mean?

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: You are right, we can ensure this is an END_OBJECT here I think

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: I meant this:

```
} else {
  //here check first that MapperService.isMetadataField(currentFieldName) returns true
  //here now parse values
}
```

**Comment by cbuescher** [comment]
@javanna thanks for the review, the last commit should address most of your comments. Still unsure if or where to put the "parseValue" method shared with GetField.

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: I modified this a bit, also tried it with String manipulation but honestly didn't like it much. Even if we controll the rendering, we should still do a basic format check (e.g. first and last character are brackets, there is the separator in between etc...) and I think RegEx communicates this much cleaner. But I guess thats a matter of taste, I won't object to later changes here ;-)

Comment (tlrx) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: We could also add a static `ShardId.fromString(String)` method next to the current `ShardId.toString()` method and do the parsing job there using simple string manipulation and simple unit tests in ShardIdTests.

With the special chars escaping the current regexp is not easy to read :(

Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: Not sure if ShardId.fromString is the right way either, unless there are other places where we print it out this weird way. Are there any? I would expect to have proper fields for each part instead. Not sure why we have this [index][id] thing here. So @tlrx decision is yours, regex or not regex? :D

So

Comment (tlrx) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: I feel like we can remove the regular expressions and manipulate strings instead. I prefer to reserve reg exp for complicated cases, here it is just enclosed strings.

**Comment by cbuescher** [comment]
@tlrx I moved the shardId parsing to a ShardId#fromString() method as you suggested, mind to take another quick look at the last commit if this is what you had in mind?

**Comment by tlrx** [comment]
@cbuescher The change 025fc58 looks good to me, thanks

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: I left the value parsing here in InternalSearchHit and use that from GetFields for now.

**Comment by cbuescher** [comment]
I have rebased and squashed this a little, waiting for another clean CI run before I will merge this.

**Comment by cbuescher** [comment]
retest this please

File: core/src/main/java/org/elasticsearch/index/shard/ShardId.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/shard/ShardId.java]: s/loose/lose

File: core/src/main/java/org/elasticsearch/index/shard/ShardId.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/shard/ShardId.java]: s/we best/the best

File: core/src/main/java/org/elasticsearch/index/shard/ShardId.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/shard/ShardId.java]: remove?

File: core/src/main/java/org/elasticsearch/search/SearchHit.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/SearchHit.java]: thanks for doing this!

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: I would make it clear that this method is about parsing stored fields values. rename it and add javadocs. I think it may also make sense to move it to XContentParserUtils, but I don't have a strong opinion.

File: core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitTests.java
Comment (javanna) [inline core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitTests.java]: It is a bit odd that we are testing here the `ShardId#fromString` method. In fact I am not even sure this method is generic enough to belong in `ShardId`...

Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitTests.java]: I followed the suggestion of @tlrx to do the shrd id parsing in "fromString" in ShardId along side of the "toString()" method. I think this is good because because the proximity clearly communicates that one belongs to the other, so I'd like to keep it there. As for the location of the test: Ideally this would be placed in a new `ShardIdTests` but that class should then have some more tests. I'd prefer to do this in a follow up PR.

Comment (javanna) [inline core/src/test/java/org/elasticsearch/search/internal/InternalSearchHitTests.java]: then create ShardIdTests and add this test to it, which is what we have. at least what we are adding is consistent then

**Comment by cbuescher** [comment]
Merged to 5.x with 7c5b85e37d0a29bd0298d4912b2af3d919f5cd15.

**Pull request merged**
